Animated particlespawners

velartrill opened the pull request:
fixes #11505, adding many new features to particlespawners and cleaning up some of the particle code while maintaining backward-compatibility with older servers and clients. with this PR, minetest particles take a step towards parity with the particle systems of other game engines, giving much more control and flexibility to mod and game devs, particularly with the new animation system, which could be extended to a full keyframe system without too much difficulty. i believe this is all in line with the "graphics improvement" part of the roadmap?

see the lua_api.txt diff for a full list of new particlespawner parameters

## To do

this PR is a ready for review i think

## caveats

i've never contributed to the minetest engine before so i had a limited understanding of where things were supposed to go, and i may have duplicated mechanisms that already existed within the codebase simply because i couldn't find them. this is also the first thing i've written in C++ in quite a few years. so there are probably a lot of organizational issues that will need addressing and i may have written parts in a insufficiently idiomatic way, sorry.

## How to test

this PR adds a very wide range of features. the ideal way to test it would be to take an existing mod that uses particles, and update its `minetest.add_particlespawner` calls to use the new parameters. the `doc/lua_api.txt` file thoroughly documents all the new additions. testing with patched clients against older servers and with patched servers against older clients to make sure particlespawners continue to behave as expected is also important.

appgurueu commented:
Closest point on a line works as follows:

Determine line vector (`a` = `end` - `start`). Determine vector from point to line start (`b` = `start` - `point`). Now, determine the length of the projection of this vector on the line vector by using the dot product: `x = dot(a, b)`. Divide this by the length of the line vector: `y = x / |a|`. Now, multiply by the line vector, and add the line start: `y * a + start`. This is the point closest to `point` on the line given by `start` and `end`. In Lua:

```lua
end, start, point = ...
line = end - start
point_to_start = start - point
pos_on_line = line:dot(point_to_start) / line:length()
closest = pos_on_line * line + start
```

velartrill commented:
@appgurueu thanks! so the question is, do we want to also support plane attractors as well as line attractors? i think i could hack together support for line attractors with this, but planes are a much thornier question. if we want them, are we going to restrict devs to planes aligned to an axis (basically a start position coupled with a vector where one of the values has to be 0), or are we going to support rotation of some kind (e.g. planes specified as four arbitrary vertices)? 🥴 i miiight be able to implement axis-locked plane attractors with some help (kind of lame but maybe it has its uses) but rotated ones are far far beyond me

cc @hecktest 

thinking about this it occurred to me that we could add support for nodebox-shaped emitters in addition to the cubes, ovoids, and roundcubes that are already supported… maybe a project for another day if anyone sees a use for that haha :)

appgurueu commented:
> @appgurueu thanks! so the question is, do we want to also support plane attractors as well as line attractors? i think i could hack together support for line attractors with this, but planes are a much thornier question. if we want them, are we going to restrict devs to planes aligned to an axis (basically a start position coupled with a vector where one of the values has to be 0), or are we going to support rotation of some kind (e.g. planes specified as four arbitrary vertices)? woozy_face i miiight be able to implement axis-locked plane attractors with some help (kind of lame but maybe it has its uses) but rotated ones are far far beyond me

Regarding planes: A plane is given by two tension vectors and a displacement vector. The vector from point to plane minus the displacement must be a multiple of the normal vector of the plane. This means that the point must equal `tension_vector_1 * x + tension_vector_2 * y + normal_vector * z`. As those three vectors are orthogonal (normal vector is just the cross product of the tension vectors, normalized), this gives us a linear system of equations solvable for `x` and `y` (for example using gaussian elimination). This should work just fine for infinite planes. If the plane is finite, `x` and `y` are constrained (usually from 0 to 1, inclusive). If you now get values for `x` and `y` which do not fulfill the constraints, the point is outside the plane bounds if projected on plane space (`x` and `y` are basically just coordinates in plane space), meaning the closest point is on one of the plane edges. You can now run the method described for lines for the four plane edges, picking the closest of the four points (this last step can probably be optimized by using `x` and `y`).

velartrill commented:
yeah i'm sorry this is all completely over my head, i don't even know most of that terminology :( i'll take a look at the article tho and see if there's anything i might be able impl myself

AFCMS commented on doc/lua_api.txt:
```diff
@@ -8571,7 +8567,7 @@ Used by `minetest.add_particlespawner`.
         -- If true particles collide with objects that are defined as
         -- `physical = true,` and `collide_with_objects = true,`.
-        -- Requires collisiondetection = true to have any effect.
+        -- Requires collisiondetection = true to have any effect.v
```
Is that change normal?

velartrill replied:
oops no that's a typo lol

nerzhul commented on src/particles.h:
```diff
@@ -71,9 +322,20 @@ struct ParticleParameters : CommonParticleParams {
+
+	std::vector<ServerParticleTexture> texpool;
+
+	ParticleParamTypes :: v3fRangeTween
```
no spaces around the double dots

velartrill replied:
@nerzhul this was fixed months ago and was trivial in any case, please approve

sfan5 commented on src/client/particles.cpp:
```diff
@@ -33,33 +33,19 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 */
 
+ClientParticleTexture::ClientParticleTexture(ParticleTexture p) :
+	ParticleTexture(p) {};
```
Generally it's a good idea to move methods without body into the class definition.

velartrill replied:
really, that doesn't cause linking issues? okay, i'll move it

sfan5 commented on src/client/particles.cpp:
```diff
@@ -577,30 +684,31 @@ void ParticleManager::addNodeParticle(IGameDef *gamedef,
 	ParticleParameters p;
-	video::ITexture *texture;
+	ClientParticleTexture texture;
+	texture.ref = nullptr;
```
Looks like a common pattern, you can avoid this by initializing `ref` in the `ClientParticleTexture` ctor

sfan5 commented on src/particles.cpp:
```diff
@@ -18,7 +18,58 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+}
+void ServerParticleTexture::deSerialize(std::istream &is, u16 protocol_ver) {
+	u8 flags = readU8(is);
+	animated = (flags |= 1<<0) > 0;
```
```suggestion
	animated = !!(flags & 1);
```
also wouldn't hurt to put them in an enum or #define

velartrill replied:
whooooops. good catch ty fixed

i didn't bother adding names to the flags simply because there's only one flag right now and they will only ever be used in two specific places, but if it'll make the protocol easier to make sense of i'll move the magic numbers into an enum

sfan5 commented on src/particles.h:
```diff
@@ -20,19 +20,269 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+
+		T blend(float fac) const {
+			// warp time coordinates in accordance w/ settings
+			if (fac > beginning) {
+				// remap for beginning offset
+				auto len = 1 - beginning;
+				fac -= beginning;
+				fac /= len;
+
+				// remap for repetitions
+				fac *= reps;
+				if (fac > 1) // poor man's modulo
+					fac -= (decltype(reps))fac;
+
+				// remap for style
+				switch (style) {
+					case TweenStyle::fwd: /* do nothing */  break;
+					case TweenStyle::rev: fac = 1.0f - fac; break;
+					case TweenStyle::pulse:
+					case TweenStyle::flicker: {
+						if (fac > 0.5f) {
+							fac = 1.f - (fac*2.f - 1.f);
+						} else {
+							fac = fac * 2;
+						}
+						if (style == TweenStyle::flicker) {
+							fac *= myrand_range(0.7f, 1.0f);
+						}
+					}
+				}
+				if (fac>1.f)
+					fac = 1.f;
+				else if (fac<0.f)
+					fac = 0.f;
+			} else {
+				fac = (style == TweenStyle::rev) ? 1.f : 0.f;
+			}
+
+			return start.interpolate(fac, end);
+		}
```
...this however should reside in a `.cpp` (rule of thumb is more than three lines)

velartrill replied:
this is a template. it cannot go in a cpp file unless it is only used in that file. since this is a type that's used in the network protocol that's a nonstarter

sfan5 replied:
Oops missed that.

sfan5 commented on src/script/lua_api/l_particleparams.h:
```diff
@@ -0,0 +1,199 @@
+	}
+
+	inline void readLuaValue(lua_State* L, TweenStyle& ret) {
+		const char* const opts[] = { "fwd", "rev", "pulse", "flicker", NULL };
+		const TweenStyle optmap[] = {
+			TweenStyle::fwd,
+			TweenStyle::rev,
+			TweenStyle::pulse,
+			TweenStyle::flicker,
+		};
+		static_assert(
+				(sizeof opts / sizeof opts[0])-1 ==
+				(sizeof optmap / sizeof optmap[0]),
+				"option maps not synced");
+
+		if (lua_isstring(L, -1)) {
+			size_t v = luaL_checkoption(L, -1, opts[0], opts);
+			ret = optmap[v];
+		}
```
we have `string_to_enum` (common/c_content.h) for this

velartrill replied:
oh thank you!

sfan5 commented on src/script/lua_api/l_particleparams.h:
```diff
@@ -0,0 +1,199 @@
+		readLegacyValue(L, name, field.start);
+
+		set_uniform: field.end = field.start;
+		done: lua_settop(L, tbl); // clean up after ourselves
```
and `StackUnroller` (common/c_types.h) for this

velartrill replied:
oh that's clever

velartrill replied:
ah, actually, no, that would not work here. apart from the `tbl` value being used as a lookup target, the branch in 185 needs to be able to roll back the stack early if the top of the stack isn't nil, and StackUnroller's top value is private without a getter for some reason

sfan5 commented on src/script/lua_api/l_particleparams.h:
```diff
@@ -0,0 +1,199 @@
+			readLuaValue(L, field.reps);
+			lua_pop(L, 1);
+
+			lua_pushliteral(L, "start"), lua_gettable(L, -2);
```
simpler: https://pgl.yoyo.org/luai/i/lua_getfield

velartrill replied:
agh right of course

sfan5 commented on src/script/lua_api/l_particles.cpp:
```diff
@@ -38,6 +39,35 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	lua_settop(L, top);
+	return;
+
+	error: lua_pushliteral(L, "invalid type in texture specifier"), lua_error(L);
```
To throw an exception you can do this
```suggestion
	throw LuaError("invalid type in texture specifier");
```

But instead of the explicit type checks I'd use `luaL_checktype`

sfan5 commented on src/util/numeric.cpp:
```diff
@@ -46,11 +46,22 @@ void myrand_bytes(void *out, size_t len)
+float myrand_float()
+{
+	u32 uv = g_pcgrand.next();
+	return (float)uv / (float)((u32)-1);
```
less confusing this way:
```suggestion
	return (float)uv / (float)U32_MAX;
```

sfan5 commented on src/util/numeric.h:
```diff
@@ -446,3 +448,18 @@ inline irr::video::SColor multiplyColorValue(const irr::video::SColor &color, fl
 }
+
+template <typename T> inline T numericAbsolute(T v) { return v < 0 ? T(-v) : v;                }
+template <typename T> inline T numericSign(T v)     { return T(v < 0 ? -1 : (v == 0 ? 0 : 1)); }
```
`std::abs` and can probably do the latter with `std::copysign`

velartrill replied:
neither of these are templates. std::abs is probably defined for enough types that we could probably get away with it (tho casting would be involved for anything smaller than an int), but numericSign is designed only for floats. i'm kind of opposed to needlessly casting numbers from integers to floats and then back again just to extract the sign

still, c++ is so picky about types (especially when it comes to resolving overloads) that i'm hesitant to use a non-template for these things

velartrill replied:
e.g., you're trying to encode the sign of an `int8_t` into a packet, and do `pkt << std::copysign(myint)` on one end, then do `readI8(is)` on the other end, and then the deserializer breaks for no apparent reason until you figure out that `myint` was being silently converted to float, and thus the sign was encoded as a float in the packet because the compiler resolved a different `operator<<` than you intended, so `readI8()` was trying to decode part of a float as an unsigned byte and leaving the rest of the float in the stream for the next function to choke on.

similar issues could happen with `std::abs` for types smaller than int or for any(!) unsigned types. there's a reason a lot of people are wary of using the STL in complex projects, and this is exactly the sort of project where i worry this kind of stuff could cause very difficult-to-find problems down the road

sfan5 replied:
I doubt it'd cause concrete problems for *us* but those are some good reasons so it's fine.

sfan5 commented on src/particles.h:
```diff
@@ -20,19 +20,269 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	bool animated = false;
+	TileAnimationParams animation;
+	ParticleParamTypes::f32Tween alpha = (f32)1;
+	ParticleParamTypes::v2fTween scale; //= (v2f){1.f,1.f};
```
FYI I think this would have worked:
`ParticleParamTypes::v2fTween scale = ParticleParamTypes::v2fTween(v2f(1.f), v2f(1.f));`
or even this perhaps
`ParticleParamTypes::v2fTween scale(1.f, 1.f);`

velartrill replied:
i believe i tried both of those and got errors

appgurueu replied:
I don't think brace syntax is a compiler extension

velartrill replied:
brace syntax is not, no. what is considered a compile-time constant expression however is largely up to compilers to define (and we don't have constexpr/consteval in c++11). my compiler and several of the compilers in your test suite accepted the brace syntax as a constant expression, a bunch of others (including the compiler used for android) did not. this was the required fix to enable the code to compile on those compilers.

sfan5 commented on src/client/particles.cpp:
```diff
@@ -251,29 +259,53 @@ ParticleSpawner::ParticleSpawner(
 	const ParticleSpawnerParameters &p,
 	u16 attached_id,
-	video::ITexture *texture,
+	ClientParticleTexture* texpool,
+	size_t texcount,
```
please use an [`std::vector`](https://en.cppreference.com/w/cpp/container/vector) instead.

velartrill replied:
is that really wise? vectors are for arrays that change size of the course of their lifetimes. `texpool` will never change size under any circumstances (unless we later decide to add a command to alter live particlespawners), but we'll end up paying for the extra space overhead anyway. also, the accessor operator (which will be called *every time a new particle is spawned*, which could be easily be thousands of times per second even on a very small server, especially iwth weather mods in use) has significant extra overhead beyond the simple add+deref that is currently used -- apart from the overhead of a function call, an extra level of indirection is necessary because `std::vector` has to account for the possibility that its contents will be moved somewhere else in memory between calls, something that will *never happen* to a texpool. and all this overhead will even affect particlespawners that only have a single texture.

sfan5 replied:
> vectors are for arrays that change size of the course of their lifetimes. texpool will never change size under any circumstances 

True, `std::array` would be a fixed size type but it has to be fixed at compile time so that won't help here.

> the accessor operator has significant extra overhead beyond the simple add+deref that is currently used

`operator[]` does no bounds-checking, you might be thinking of [`at`](https://en.cppreference.com/w/cpp/container/vector/at). Other than that it should be no different than a C-style indexing.

> apart from the overhead of a function call

Templates are inlined

> an extra level of indirection is necessary because std::vector has to account for the possibility that its contents will be moved somewhere else in memory between calls

I don't know what this is referring to and this is my first time hearing it. A vector internally encapsulates a pointer to a continuous memory region containing its contents, I don't see the difference here: it could be swapped out just like a pointer kept manually.

velartrill replied:
update: okay, i ran some tests. looks like `std::vector` (at least on g++) is nowhere near as bad as it was when i was 17, but there's still a nearly 2x slowdown compared to normal array accesses:

```c++
$ cat vectest.cc
#include <vector>
#include <iostream>

int main(int argc, char** argv) {
	std::vector<int> vec = {1,2,3,4,5,6,7,8,9,10,11};
	int* arptr = new int[11]{1,2,3,4,5,6,7,8,9,10,11};
	size_t q = 0;
	for (size_t i = 0; i < 1ULL<<32; ++i)
#		ifdef test_vector
			q += vec[argc];
#		else
			q += arptr[argc];
#		endif

	return q;
}

$ g++ vectest.cc -ovectest -Dtest_vector
$ g++ vectest.cc -optrtest
$ time ./ptrtest 1

________________________________________________________
Executed in    7.04 secs   fish           external 
   usr time    7.03 secs  798.00 micros    7.03 secs 
   sys time    0.00 secs  214.00 micros    0.00 secs 

$ time ./vectest 1

________________________________________________________
Executed in   12.31 secs   fish           external 
   usr time   12.31 secs    0.00 micros   12.31 secs 
   sys time    0.00 secs  1079.00 micros    0.00 secs 
```

(i also tried compiling the tests with -O4 but g++ is clearly outsmarting me in that case, because ever absurd numbers like `1LL<<63` finish immediately under both implementations; i looked at them with a disassembler and there's no loop at all in those versions)

sfan5 replied:
Hmm that's unexpected to me.
Testing a different example that can't be optimized away (a sum) I found a case where assembly for `std::vector` and `int*` are exactly identical: https://godbolt.org/z/c8bazY67o

> (i also tried compiling the tests with -O4 but g++ is clearly outsmarting me in that case, because ever absurd numbers like 1LL<<63 finish immediately under both implementations; i looked at them with a disassembler and there's no loop at all in those versions)

I tested that and even with `-O2` both gcc and clang remove(?) the loop here.

　

In any case we use vectors and other C++ containers a lot since they're less error prone and more convenient to work with than pointers, so if the overhead (which exists depending on how you use them) breaks performance on someone's machine I bet they weren't able to run Minetest anyway.

velartrill replied:
> Hmm that's unexpected to me.

take a look at `echo "#include <vector>" | g++ -xc++ -E -P` and you'll probably get a general sense at least of why this might be haha (if you can dig through all the abstraction and template magic; i got lost somewhere around `_Vector_impl_data`)

> Testing a different example that can't be optimized away (a sum) I found a case where assembly for std::vector and int* are exactly identical: https://godbolt.org/z/c8bazY67o

it's interesting but my test shows that the problem *can* occur, which is what's important; it's unwise to rely too heavily on compiler optimizations as an excuse to not just do the right thing in the first place. minetest is much much muuuch more complex than these little examples, and we wouldn't be able to know to what degree a given compiler could optimize away the vector baggage without digging into the assembly of each one (which, ouch) or how those optimizations might change in future versions of the compiler (they're not part of the c++11 standard, after all). and here's the thing: even in the most optimistic case, where the vector access is boiled down to exactly the same code as the array access, all it takes is one unsuspecting programmer a few years down the road making a call to some particular vector function that arrays don't have (say `push_back()`) to potentially cause an explosion of complexity by invalidating optimisations we were relying on; can you imagine trying to track down something like that?

so i'm really sorry but i just don't think it's acceptable to further worsen particle performance (even if only under certain compilers/stdlibs!) which, as hecktest has pointed out, is already pretty bad, just for a mildly convenient new feature (it would be different if these were necessary for something very important of course), especially as you seemingly (apologies if i've misunderstood or missed something) have no specific reason for wanting to use vectors in a performance-critical part of the code and vector features are of no use here. it's a bit like using chromium instead of `cat` to read to a text file (huge exaggeration ofc, but you get my drift). "we can probably get away with it" is a far cry from "we need to," and having succumbed to that mindset on many occasions, i have learned the hard way that these things add up -- and none of those projects were anywhere near the size and complexity of minetest. it's easier to speed up code when there's ten obvious inefficencies, much much harder when there's a thousand tiny little inefficiencies that would all be fine on their own but together bring a system to its knees, especially because they're so damn hard to find once they've been written.

(for example: i noticed last night that `ParticleManager` does not preallocate space (i.e. with `reserve`) when a new particlemanager is added! this completely escaped my notice until now, even tho i've been tits-deep in the particle engine code off and on for two weeks. i've started on a fix but it needs a tiny bit of refinement to handle edge cases before it can be merged)

bottom line: if you have a *concrete* objection to pointers (as you yourself have indicated, vectors do not perform bounds checking, so they are no less error-prone than pointers in cases where the arrays do not change over their lifetime), i would very much like to work with you to figure out an optimal solution! for instance, if your concern is that raw malloc'd arrays don't cooperate with the RAII idiom that this project makes extensive use of (which is a very reasonable objection! i did have to add a destructor that wasn't there before to accomodate this array), a simple `flexArray<T>` template that behaves like `std::arrray<T,N>` but uses heap storage and allows arbitrary sizes would be a perfect compromise; it would address the actual problem without impacting performance in the slightest.

if that's not good enough for you, or this is just about discomfort with manual array use rather than a concrete problem arrays introduce, then tbh i believe the right thing to do is to rip out texpool support and just return to one texture per emitter. if use of vectors had been the only way to implement this functionality, i would never have added it in the first place. rubenwardy's initial reaction to #11505 was wise and very much warranted -- i don't think i should be cramming in new features that hit performance-critical areas, even slightly, without extremely good reason; i already killed particle texture blending over this and this isn't nearly as useful. so i'm not interested in being responsible for further slowing the engine down over it, not even mildly. especially since adding actually important features has already required non-negligible structure size increases (affecting both the network protocol and local memory on client and server) and further complications of code that has to run many times each second on the client. a visual equivalent to texpool support could be implemented purely in a couple lines of lua if we wanted to; there's no real reason it *has* to be in the engine… and perhaps most importantly, i'd rather focus my attention on the features that motivated me to make this PR in the first place haha :)

sfan5 replied:
Well it doesn't look like we're coming to an agreement here. Essentially it's a matter of whether it is premature optimization or not, I believe it is and continue to insist that `std::vector` is the better alternative.

If other devs chime in a decision could be made but that can wait.

> particle performance which, as hecktest has pointed out, is already pretty bad

His remark was regarding the rendering performance of particles, which suffers because vertices are constantly being modified. You didn't add more of that so don't worry.

> if [...], then tbh i believe the right thing to do is to rip out texpool support and just return to one texture per emitter.

I don't think so. Even if texture pools can be replicated by creating multiple particle spawners with differing parameters from Lua it's still an useful feature IMO.

> i'd rather focus my attention on the features that motivated me to make this PR in the first place haha :)

Please do.

velartrill replied:
update: hopefully this will be an acceptable compromise? i've removed the raw array and switched to an `std::unique_ptr`, this provides the benefits of using a vector without most of the unnecessary baggage or capabilities that aren't needed

velartrill replied:
> If other devs chime in a decision could be made but that can wait.

alright. i'll leave the code in place pending; hopefully we can come to an agreement on a more appropriate solution than `std::vector`. failing that, i ask that this PR not be merged until i've had time to remove the texpool code; i hope you can understand why i'm not interested in bearing the `git praise` for creating what i consider to be needless technical debt.

sfan5 commented on src/particles.h:
```diff
@@ -20,19 +20,269 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+		static This pick(float* f, const This& a, const This& b) {
+			return This(vectorBlend(f, a.val, b.val));
+		}
+	};
```
Although surely convenient I'm not sure if I'm a fan of creating these thin wrappers around scalar/vector types...

velartrill replied:
alright, i've replaced this with overloaded functions

sfan5 commented on builtin/game/features.lua:
```diff
@@ -21,6 +21,7 @@ core.features = {
 	use_texture_alpha_string_modes = true,
 	degrotate_240_steps = true,
 	abm_min_max_y = true,
+	particlespawner_tweenable = true
```
Missing documentation in lua_api.txt

sfan5 reviewed (commented):
Had a look at the code to give some mostly general advice / feedback

appgurueu commented on doc/lua_api.txt:
```diff
@@ -8603,8 +8602,221 @@ Used by `minetest.add_particlespawner`.
+        -- old syntax
+
+        pos = 0;
+        -- all components of every particle's position vector will be set to this
```
> tbh I don't think forcing modders to write out vectors is so bad that this should be supported, other devs may disagree

I agree. If anything, there should be a new vector utility constructor, but that's for a different PR.

velartrill replied:
there's a specific reason i allowed for this syntax; the `pos` one is just for sake of example and not very useful. however, there are some properties, like radius and drag, that in the majority of situations you're only ever going to want a unifom value. however, there are rarer cases when e.g. defining drag as a vector is useful, say to avoid interfering with gravity applied by acc. i didn't think it was reasonable to force the extra syntactic salt on the majority of modders just to give an extra feature to a minority, but if the consensus is that this is uncool i'll rip it out?

velartrill replied:
so, is the consensus that i should remove this?

sfan5 replied:
Hard to say without at least one other coredev having looked at it.
Removing it in the end after the rest is all ready wouldn't be a problem, would it?

velartrill replied:
don't think so, just means removing a few lines of code in the lua api and changing the docs

velartrill commented:
can't respond at length rn bc running errands, but tbqh the specifics are academic -- when i was learning c++ i once made the mistake of storing image pixel data in a vector rather than a bare array, and when i tried to run the program on a machine that wasn't my souped-up desktop the frame rate slowed to an unusable crawl (the data was being redrawn each frame). doing nothing more than replacing the vector with an array brought it up to 40-50fps, an order of magnitude faster than it had been before. the performance loss is substantial.

what specifically are your concerns that make you want to use a vector here? i'm sure we can find an alternative that addresses them without sacrificing further performance 

On Friday, August 20, 2021, sfan5 wrote:
> @sfan5 commented on this pull request.
> 
> 
> 
> > +	ClientParticleTexture* texpool,
> +	size_t texcount,
> 
> > vectors are for arrays that change size of the course of their lifetimes. texpool will never change size under any circumstances 
> 
> True, `std::array` would be a fixed size type but it has to be fixed at compile time so that won't help here.
> 
> > the accessor operator has significant extra overhead beyond the simple add+deref that is currently used
> 
> `operator[]` does no bounds-checking, you might be thinking of [`at`](https://en.cppreference.com/w/cpp/container/vector/at). Other than that it should be no different than a C-style indexing.
> 
> > apart from the overhead of a function call
> 
> Templates are inlined
> 
> > an extra level of indirection is necessary because std::vector has to account for the possibility that its contents will be moved somewhere else in memory between calls
> 
> I don't know what this is referring to and this is my first time hearing it. A vector internally encapsulates a pointer to a continuous memory region containing its contents, I don't see the difference here: it could be swapped out just like a pointer kept manually.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/minetest/minetest/pull/11545#discussion_r69293702

velartrill commented:
so, re the protocol bump -- looking over the current impl i'm not actually totally sure it's strictly necessary. it might be possible to tweak things so a bump isn't necessary, depending on how exactly serialization works, which i'm not yet clear on; right now i think the only thing the protocol version is being checked for is in `network/clientpackerhandler.cpp:1024` where we check if the version >= 41 before attempting to deserialize the new fields

sfan5 commented:
If you can avoid the protocol bump without the (de)serialization functions looking like a mess that's fine, if not that's also fine.

velartrill commented:
okay, i should be able to avoid the protocol bump then i think!

appgurueu commented:
Do we want finite planes or lines? In the infinite case, finding the closest point is actually pretty simple.

AFCMS commented on doc/lua_api.txt:
```diff
@@ -8603,8 +8607,221 @@ Used by `minetest.add_particlespawner`.
+        -- visible, at 0, it is completely invisible.
+
+	    scale = {x = 0, y = 0};
+	    -- scales the texture onscreen
```
Is that indentation correct? It doesn't look consistent

velartrill commented:
this PR is now feature-complete (pending last-minute requests) and the documentation has been updated. as far as i'm concerned it is ready to merge in its current state.

the API i settled on for attractor objects is as follows:

```lua
attract = {
    kind = 'none' / 'point' / 'line' / 'plane';
    origin = vector.new(0,0,0); -- center for 'point'; point on line or plane for others
    origin_attached = minetest.getPlayerByName(player); -- allows origin to be specified as an offset from an object
    angle = vector.new(1,1,1); -- a second point on the line, or the normal of the plane; not used for 'point'
    angle_attached = (…); -- like origin_attached, but for angle
};
```
ping @AFCMS

sfan5 commented on doc/lua_api.txt:
```diff
@@ -8680,10 +8685,22 @@ Used by `minetest.add_particle`.
 
 `ParticleSpawner` definition
```
In the section for individual particles (above here), is `drag` missing? I saw it in some code and am wondering if you added support for it to add_particle too.

velartrill replied:
i don't think i did, i'll see about adding that

velartrill replied:
`drag` and `bounce` can now be set from `add_particle`

sfan5 commented on doc/lua_api.txt:
```diff
@@ -8751,8 +8755,258 @@ Used by `minetest.add_particlespawner`.
     }
 
+### Modern definition fields
+After v5.5.0, spawner properties can be defined in several different ways depending
```
The places were it says 5.5.0 will have to be changed and there's a few code style changes here and there but I won't burden you with that, me (or someone else) can do it before the merge.

velartrill replied:
i'll fix the 5.5.0 stuff where i find it at least

velartrill commented on doc/lua_api.txt:
```diff
@@ -8751,8 +8755,258 @@ Used by `minetest.add_particlespawner`.
+        img = 'mymod_particle_texture.png';
+        -- the texture specification string
+
+        alpha = 1.0;
```
`alpha_tween` and `scale_tween` are also here yeah, and yes, it does. all the tweenable properties are "parsed" by the same function. i'll update the docs to make this clearer

velartrill commented on doc/lua_api.txt:
```diff
@@ -8751,8 +8755,258 @@ Used by `minetest.add_particlespawner`.
+
+    texpool = {
+        'mymod_particle_texture.png';
+        { img = 'mymod_spark.png', fade = 'out' };
```
seems to be. i remember little about this but i think this was a primitive mechanism i replaced by adding more precise controls to the tween tables, `{start, style, reps}`. in current syntax this would be expressed:

```
texpool = {
    {name='mymod_spark.png', alpha_tween={1,0, style='fwd'}} -- 'fwd' is the default
}
```

velartrill commented on doc/lua_api.txt:
```diff
@@ -8751,8 +8755,258 @@ Used by `minetest.add_particlespawner`.
+    the position of an entity rather than a coordinate in world space.
+  * vec3 `angle_attached`: allows the angle to be specified as an offset from
+    the position of an entity rather than a coordinate in world space.
+  * float range `strength`: the speed with which particles will move towards
```
for strength? `0` is the correct default. it is specifying a speed in nodes per second and is not constrained to [0,1], so `1` makes no more sense than `2` or `69`. i'll make it more prominent in the docs tho

velartrill replied:
also the docs were formatted wrong lol, i put that in the wrong section

velartrill commented:
@appgurueu if it's that big a deal i'll see if i can come up with a regex to automate the change without wrecking any of the prose. otherwise it's not worth all the time it would take since those forms are semantically identical

appgurueu commented:
> @appgurueu if it's that big a deal i'll see if i can come up with a regex to automate the change without wrecking any of the prose. otherwise it's not worth all the time it would take since those forms are semantically identical

I can do the conversion to save you the time, as I'm the one caring about minor inconsistencies here.

velartrill commented:
> I made some time today to finally test this PR since you've been waiting long enough. It works great. My tests were far from exhaustive but I found a few things that didn't function:
> 
> * setting `bias` of a min/max range doesn't do anything at all
> * I couldn't get `alpha_tween` to work, this might also be a misunderstanding of the docs
> 
> The test mod source is here: https://gist.github.com/sfan5/f3b8c427c5873474c947df20ff689e22 you can find the mentioned cases with ctrl+f `-- bug`

```
texture = {
	img = "default_grass_3.png",
	alpha_tween = {1, 0, style = 'fwd'}, -- BUG neither of these work
},
```

this is the correct usage (alpha is a property of the texture, not the particle), i have no idea why it wouldn't be working bc it worked fine the last time i tested it. i'll give it another shot, maybe i did something later that broke it somehow.

i am absolutely stumped by the problem with `bias`; i may need some help with this one. i went and graphed the equations in use and `qalculate` at least is showing absolutely the correct (intended) values from those expressions

velartrill commented:
@appgurueu thanks, i would appreciate that

indentation stuff is not intended, i'm using kate and it likes screwing up my indents with no warning. will fix

velartrill commented:
@sfan5 i cannot reproduce your problem with alpha tweening. i tested by tweaking some code from my mod; i attached the snippet of code that worked for me for reference, but really only the texpool fragment should be relevant

```
return minetest.add_particlespawner {
	amount = 450;
	time = 2;
	-- old syntax
	minpos = vector.subtract(tgt.under, 0.5);
	maxpos = vector.add(tgt.under, 0.5);
	minvel = minvel, maxvel = maxvel;
	minexptime = 1, maxexptime = 2;
	minsize = 0.5, maxsize = 2;
	texture = L.image('sorcery_spark.png'):multiply(color):render(); -- for legacy clients
	texpool = {
		{
			name = 'sorcery_inferno_crystal.png'; -- static image
			alpha_tween = { 1, 0; style = 'fwd'};
			scale_tween = { {x=2,y=3}; {x=0,y=0}; style = 'pulse', reps = 6 };
		};
	};
}
```

maybe this is a problem with the image format somehow? some issue with how irrlicht deals with certain kinds of pngs? i am really baffled

velartrill commented:
@sfan5 fixed the `alpha_tween` bug, it should work from both `texture` and `texpool` now. still stumped by why `bias` isn't working

velartrill commented on doc/lua_api.txt:
```diff
@@ -8751,8 +8755,258 @@ Used by `minetest.add_particlespawner`.
+            -- number of times the animation is played over the particle's lifespan
+
+            start = 0.0;
+            -- point in the particle's lifespan at which the animation begins. 0 is
```
fixed

appgurueu commented:
Here's the announced PR for the consistency: https://github.com/velartrill/minetest/pull/1

While working on it, I realized the docs were already inconsistently using single quotes sometimes, so I fixed that too: https://github.com/minetest/minetest/pull/12075/files

I could indeed use a RegEx to find most occurrences, but ended up having to edit them manually.

sfan5 commented:
Just gave this another test, unfortunately I don't get anything other than
`2022-02-18 21:07:30: ERROR[Main]: ServerError: AsyncErr: Lua: Runtime error from mod 'test2' in callback LuaABM::trigger(): <no description>`
regardless of how the particlespawner is created. (updated the gist with my code, but not much has changed)
It's also weird that no traceback is shown, are you throwing `LuaError`s in the glue code?

As for the use of `pow`: I roughly extracted the relevant code and the values returned by [my test program](https://gist.github.com/sfan5/f3b8c427c5873474c947df20ff689e22#file-99-test-cpp) are definitely fine. So the bug is either in the template stuff I didn't reproduce or somewhere else entirely (do the values arrive at the client correctly?) ¯\\\_(ツ)_/¯

velartrill commented:
the error's probably bc you forgot to update `img` → `name` :) that should have a more useful error tho

sfan5 commented:
Ahh right, totally missed that.
`luaL_checkstring` is the correct function, I don't get why an error message is missing.

sfan5 commented:
Figured out the `bias` issue, the value got lost on two occasions. I'll send a PR with changes.

velartrill commented:
…_wow_ do i hate C++. i am so glad you caught that bc im not sure i ever would have. thank you!!

sfan5 reviewed (approved):
Tested, works.

lhofhansl commented:
@velartrill Thanks for taking this on.
Can we add a sample spawner to devtest?

velartrill commented:
go ahead and submit a PR, i'll merge it. alternatively, if you're asking *me* to add a test spawner, i'd need to know what you expect from it. i added enough new functionality to create a combinatoric explosion so testing every possible combination of new features is not going to be viable, and not all features make sense to use together in any case.

lhofhansl commented:
Sorry... Was asking as part of this. I.e. a particle spawner exercising some of the new features. And maybe you are right the combinatiorial explosion is too big to represent easily.

appgurueu commented:
> Sorry... Was asking as part of this. I.e. a particle spawner exercising some of the new features. And maybe you are right the combinatiorial explosion is too big to represent easily.

IMO, just have at least one test per feature; not every combination of features needs to be tested. Perhaps just find a way to "fuzz" combinations (that is, generate random combinations), fuzz a thousand spawners and call it a day.

velartrill commented:
i'm not sure a nondeterministic test really qualifies as a "test" :) i'll see about trying to come up with something useful if and when i have time, however tbh this is not a priority for me

TurkeyMcMac commented:
The RNGs provided by Minetest are deterministic, right?

appgurueu commented:
> The RNGs provided by Minetest are deterministic, right?

Yes. Minetest does in fact seed Lua's random BTW. Using PseudoRandom might be more useful for reproducible tests.

velartrill commented:
> > The RNGs provided by Minetest are deterministic, right?
> 
> Yes. Minetest does in fact seed Lua's random BTW. Using PseudoRandom might be more useful for reproducible tests.

recent versions of lua do set an initial seed but… it uses a combination iirc of `time(NULL)` and the integer value of a pointer to a particular object in memory. not really a mature approach. (i assume minetest uses the system entropy source?)

velartrill commented:
more on topic: i was thinking about particle effect tests and i realized two things are still missing that are kind of important for a particle system. if no one objects, i'd like to also (try to) add support for brownian motion (particle jitter) and alternate blending modes (at the very least add or screen). currently i think the only way to emulate some form of brownian motion would be by horribly abusing attractor entities

appgurueu commented:
> recent versions of lua do set a "random" initial seed but… it uses a combination iirc of `time(NULL)` and the integer value of a pointer to a particular object in memory. not really a mature approach. (i assume minetest uses the system entropy source?)

Minetest uses just the time (in seconds IIRC, meaning you could get the same sequence if starting MT the same second) for seeding Lua's `math.random`, although it does provide a `SecureRandom` that uses the system entropy source if needed.

appgurueu commented:
> alternate blending modes (at the very least add or screen).

Like fixing current blending bugs? Consider having a look at #10398.

Should such blending only apply to particle-particle or also particle-node or particle-entity rendering?

velartrill commented:
@appgurueu that issue is unrelated. i asked earlier if i should incorporate that fix into this PR but i didn't ever get an answer, so i left it alone. i would _really_ like to incorporate that PR into this one, however, if that's acceptable.

what i am talking about wrt blending is which function will be used to blit the particles onto the screen. see https://www.andersriggelsen.dk/glblendfunc.php for some examples, but screen and add should be familiar to anyone who does vfx or image editing -- these are commonly used for creating light effects, and plain alpha blending without add or screen tends to look unnatural for luminous objects (e.g. magical sparkles, fireballs, or any other effect representing some kind of energy, rather than smoke or shards or droplets of some physical material). the particle clash bug is iirc caused by problems with the stencil buffer, and won't be affected by changing the blend mode

> Should such blending only apply to particle-particle or also particle-node or particle-entity rendering?

in opengl the blend function is used wherever pixels are blitted on top of another. this means that the particles will blend with whatever is underneath them, regardless of its logical origin. for more fine-grained control you would need to render to separate textures and then blit them together. i'm not going to do this because a) it's a huge effort and b) it looks extremely unnatural to blend with some objects but not others; i cannot think of a realistic use case for this

sfan5 commented:
If samples are needed my test stuff that I linked earlier could also be adapted and put into devtest.

velartrill commented:
this is quite weird. naming `std::enable_if_t` (shorthand for `std::enable_if<…>::type` added in C++14) from `/src/particles.h` causes the compile to fail with the complaint that no such template is defined (i made sure `#include <type_traits>` was there), but it works fine from `/src/particles.cpp`. also this doesn't appear to be a standards issue, bc i was able to define my own `std::enable_if_t` equivalent in /src/particles.h just fine

unrelated: what exactly is `l_particles_local.cpp` for? i've mostly been duplicating my changes across both `l_particles{,_local}.cpp` but i noticed the latter doesn't seem to support object references, e.g. to attach particlespawners to objects. should i leave out code that looks up object IDs from that file, like for `origin_attached`?

velartrill commented:
so i went ahead and added blend mode and jitter, and incorporated the fixes from #10398, since nobody seemed to have any objections. if anyone has any problems with any of these changes, just let me know and we can revert them.

velartrill commented:
okay actually this is horribly bugged, going to try and find the problem and fix it. currently it's crashing when node particles are added for some reason

velartrill commented:
alright that should do it

x2048 commented:
@velartrill

> unrelated: what exactly is l_particles_local.cpp for?

This is particle support for client-side mods, see [client_lua_api.txt](https://github.com/minetest/minetest/blob/master/doc/client_lua_api.txt#L858)

> so i went ahead and added blend mode (currently just alpha blend, add, subtract, and a rough approximation of screen, but i left a spare bit in the protocol for adding more later) and jitter,

:+1: 

> incorporated the fixes from #10398, since nobody seemed to have any objections

Can you check whether you observe performance drop now that the fix is incorporated? It was the initial concern on #10398.

*Note*: the branch has conflicts and needs a manual rebase.

velartrill commented:
@x2048 thanks for the info!

i've been playing & hacking on mt with the #10398 patch incorporated in my personal build for several months now and never noticed any performance issues at all (i use a mid-range business laptop with decent integrated graphics), certainly not anything comparable with setting shadows on low or higher.

x2048 reviewed (changes requested):
@velartrill I have started reviewing testing the code and ran into a few issues, plus I'd like to propose corrections to the API and documentation. I have not finished yet, but could you take a look at the comments?

velartrill commented on src/client/particles.cpp:
```diff
@@ -59,25 +42,70 @@ Particle::Particle(
+			blendop = video::EBO_REVSUBTRACT;
+		break;
+
+		case ParticleParamTypes::BlendMode::screen:
```
thanks!

velartrill commented on doc/lua_api.txt:
```diff
@@ -8793,8 +8806,288 @@ Used by `minetest.add_particlespawner`.
+                type = "vertical_frames",
+                aspect_w = 16, aspect_h = 16,
+                length = -3,
+                -- repeat the animation three times over the particle's lifetime
```
correct yes

velartrill commented on doc/lua_api.txt:
```diff
@@ -8793,8 +8806,288 @@ Used by `minetest.add_particlespawner`.
+
+        alpha_tween = {1, 0},
+        -- can be used instead of `alpha` to animate the alpha value over the
+        -- particle's lifetime. these tween tables work identically to the tween
```
range values and tween tables are both defined in the previous section, which begins on 8887 "Modern definition fields"

velartrill commented on doc/lua_api.txt:
```diff
@@ -8793,8 +8806,288 @@ Used by `minetest.add_particlespawner`.
+    that they will die as they cross the attractor threshold. this behavior
+    is the default but is undesirable for some kinds of animations; set it to
+    false to allow particles to live out their natural lives.
+* vec3 range `radius`: if set, particles will be arranged in a sphere around
```
the section beginning at line 8887

x2048 replied:
I have re-read the section once again, and have some suggestions on how to improve the readability. I'll make comments to that section.

velartrill commented on doc/lua_api.txt:
```diff
@@ -8793,8 +8806,288 @@ Used by `minetest.add_particlespawner`.
+    create a vertical line that passes through `origin`. for planes, `angle`
+    is the surface normal of an infinite plane on whose surface `origin` is
+    a point. functions as an offset if `angle_attached` is also set.
+  * entity `origin_attached`: allows the origin to be specified as an offset
```
alright, i'll change it

velartrill replied:
actually, i see now why i used `attached`; using `attach` or `attach_to` creates an inconsistency with the preexisting `attached` property on the particlespawner (whose name can't be changed without breaking existing mods, tho we could add an alias i suppose). i don't think a more "intuitive" name is worth creating inconsistency with the existing API so i'm reverting my change

velartrill commented on src/particles.cpp:
```diff
@@ -37,6 +133,20 @@ void ParticleParameters::serialize(std::ostream &os, u16 protocol_ver) const
 	writeU8(os, node_tile);
+	writeV3F32(os, drag);
+	jitter.serialize(os);
+	bounce.serialize(os);
```
textures are serialized in server.cpp at 1667 (legacy texture) and 1686 (texpools)

velartrill replied:
in regards to blend parameters not working, it sounds like that's probably a problem with your GPU or how irrlicht is talking to your GPU, because i've had no problems testing blend effects on mine (AMD Renoir) so far.

velartrill commented on src/particles.cpp:
```diff
@@ -47,17 +157,20 @@ void ParticleParameters::deSerialize(std::istream &is, u16 protocol_ver)
+	if (streamEndsBeforeParam<v3f, readV3F32>(drag, is))
+		return;
+	jitter.deSerialize(is);
+	bounce.deSerialize(is);
```
textures are deserialized in `network/clientpackethandler.cpp` at 1046 (legacy texture) and 1071 (texpools)

velartrill commented on src/client/particles.cpp:
```diff
@@ -238,7 +332,8 @@ void Particle::updateVertices()
 		m_box.addInternalPoint(vertex.Pos);
-		vertex.Pos += m_pos*BS - intToFloat(camera_offset, BS);
+		// see #10398
+		// vertex.Pos += m_pos*BS - intToFloat(camera_offset, BS);
```
done

velartrill commented on doc/lua_api.txt:
```diff
@@ -8793,8 +8806,288 @@ Used by `minetest.add_particlespawner`.
+        -- **only two frames**, the first and the last, are used. any of the above
+        -- definition styles can be used here as well in any combination
+
+        pos_tween = {
```
this *is* the section about tween tables. this is where they are defined. i'm not clear on what your issue is

velartrill commented on doc/lua_api.txt:
```diff
@@ -8793,8 +8806,288 @@ Used by `minetest.add_particlespawner`.
+  * vec3 `origin`: the origin point of the shape towards which particles will
+    initially be oriented. functions as an offset if `origin_attached` is also
+    set.
+  * vec3 `angle`: sets the direction in which the attractor shape faces. for
```
renamed

velartrill commented on src/client/particles.cpp:
```diff
@@ -134,24 +175,54 @@ void Particle::render()
+	// apply drag (not handled by collisionMoveSimple) and brownian motion
+	v3f av = vecAbsolute(m_velocity);
+	av -= av * (m_drag * dtime);
+	m_velocity = av*vecSign(m_velocity) + v3f(m_jitter.pickWithin())*dtime;
```
that's not possible. 222 is an else clause, drag needs to be applied unconditionally

velartrill commented:
aaaaaghhhh i am so done editing the docs using Kate

will fix >_<

x2048 commented on src/client/particles.cpp:
```diff
@@ -59,25 +42,70 @@ Particle::Particle(
+			blendop = video::EBO_ADD;
+		break;
+
+		case ParticleParamTypes::BlendMode::sub:
```
OK, I'll so some more testing. Have you seen "sub" working?

x2048 commented on doc/lua_api.txt:
```diff
@@ -8930,8 +8943,307 @@ Used by `minetest.add_particlespawner`.
+and all lower-precedence fields will be ignored, allowing for graceful
+degradation in older clients).
+
+    {
```
@velartrill 

I overlooked these definitions because the syntax is the same as for object definitions in other sections, while you describe different data types accepted by the property.

I propose making the data types more prominent and make the following changes:
* Use `vector` instead of `vec3` because this is the only section that introduces `vec3` name.
* Declare `range` data type which can be a `vector range` and a `float range`.

For example, the section could look like this:

----

The following definitions are all equivalent, listed in order of precedence from
lowest (the legacy syntax) to highest (tween tables). If multiple forms of a
property definition are present, the highest-precidence form will be selected
and all lower-precedence fields will be ignored, allowing for graceful
degradation in older clients).

1. Old Syntax:

```
      maxpos = {x = 0, y = 0, z = 0},
      minpos = {x = 0, y = 0, z = 0},
```

2. Scalar value:

```
      pos = 0,
```
All components of every particle's position vector will be set to this value.

3. Vector:

```
      pos = vector.new(0,0,0),
```

All particles will appear at this exact position throughout the lifetime of the particlespawner.

5. Range

Ranges can be scalar (float) ranges and vector ranges. The following example shows a vector range:

```
      pos = {
            -- the particle will appear at a position that is picked at random from
            -- within a cubic range

            min = vector.new(0,0,0),
            -- `min` is the minimum value this property will be set to in particles
            -- spawned by the generator

            max = vector.new(0,0,0),
            -- `max` is the minimum value this property will be set to in particles
            -- spawned by the generator

            bias = 0,
            -- when `bias` is 0, all random values are exactly as likely as any
            -- other. when it is positive, the higher it is, the more likely values
            -- will appear towards the minimum end of the allowed spectrum. when
            -- it is negative, the lower it is, the more likely values will appear
            -- towards the maximum end of the allowed spectrum. the curve is
            -- exponential and there is no particular maximum or minimum value
        },
```

6.  Tween table:

A tween table should consist of a list of frames in the same form as the
untweened pos property above, which the engine will interpolate between,
and optionally a number of properties that control how the interpolation
takes place. currently **only two frames**, the first and the last, are
used, but extra frames are accepted for the sake of forward compatibility.
Any of the above definition styles can be used here as well in any combination
supported by the property type.

You need to append `_tween` suffix to the property name to assign a tween table.

```
        pos_tween = {
            style = "fwd",
            -- linear animation from first to last frame (default)
            style = "rev",
            -- linear animation from last to first frame
            style = "pulse",
            -- linear animation from first to last then back to first again
            style = "flicker",
            -- like "pulse", but slightly randomized to add a bit of stutter

            reps = 1,
            -- number of times the animation is played over the particle's lifespan

            start = 0.0,
            -- point in the spawner's lifespan at which the animation begins. 0 is
            -- the very beginning, 1 is the very end

            -- frames can be defined in a number of different ways, depending on the
            -- underlying type of the property. for now, all but the first and last
            -- frame are ignored

            -- frames

                -- floats
                0, 0,

                -- vec3s
                vector.new(0,0,0),
                vector.new(0,0,0),

                -- vec3 ranges
                { min = vector.new(0,0,0), max = vector.new(0,0,0), bias = 0 },
                { min = vector.new(0,0,0), max = vector.new(0,0,0), bias = 0 },

                -- mixed
                0, { min = vector.new(0,0,0), max = vector.new(0,0,0), bias = 0 },
        }
```

x2048 commented on doc/lua_api.txt:
```diff
@@ -8930,8 +8943,307 @@ Used by `minetest.add_particlespawner`.
+  velocity values within a range. the velocity calculated by this method will
+  be **added** to that specified by `vel` if `vel` is also set, so in most
+  cases **`vel` should be set to 0**. `attract` has the fields:
+  * string `kind`: selects the kind of shape towards which the particles will
```
Here I propose to go with an code snippet instead of nested lists to improve readability and be consistent with other table definitions:

```
{

  kind = "point",  -- specifies the kind of shape towards which the particles will be oriented
  -- * "none":  no attractor is set and the `attractor` table is ignored
  -- * "point": the particles are attracted to a specific point in space.
  --            use this also if you want a sphere-like effect, in combination with
  --            the `radius` property.
  -- * "line":  the particles are attracted to an (infinite) line passing
  --            through the points `origin` and `angle`. use this for e.g. beacon
  --            effects, energy beam effects, etc.
  -- * "plane": the particles are attracted to an (infinite) plane on whose
  --            surface `origin` designates a point in world coordinate space. use this
  --            for e.g. particles entering or emerging from a portal.
  
  strength = { ... float range }, -- the speed with which particles will move towards the attractor.
  -- If negative, the particles will instead move away from that object.
  
  ...
}
```

velartrill replied:
can you submit a PR with the changes you want to make to the documentation? i would like to avoid a drawn-out back-and-forth over this if possible :)

x2048 commented:
@velartrill Could you take a look at velartrill/minetest#3?

sfan5 commented:
@x2048 is this okay from your side now?

x2048 reviewed (commented):
I have submitted proposed corrections in velartrill/minetest#4 and will approve as soon as the corrections are applied.

@velartrill @sfan5 @rubenwardy

sorcerykid commented:
This is frankly one of the most elaborate APIs I've seen in the past few years for Minetest. The level of detail in the documentation is also appreciated. You put an immense amount of effort into the PR, so I'm glad to see it merged in time for the upcoming feature freeze. Thank you for all your hard work!

velartrill commented:
@sorcerykid it's feedback like this that makes contributing to FOSS worthwhile :) you are very welcome!

appgurueu commented:
I'll definitely use the tweens for alpha & size to fade out my particles from now on; before this PR you would've had to abuse animations for this (and couldn't have the spawner spawn particles with variable lifetimes since the animation would then not have matched the particle lifetime).

andriyndev commented:
https://github.com/luanti-org/luanti/issues/16724
